compiler/optimizer/LoopVersioner.cpp

```diff
@@ -4775,6 +4815,22 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
       properRegion->removeExternalEdgeTo(predBlock->getStructureOf(), succBlock->getStructureOf()->getNumber());
       }
 
+   if (seenOSRGuards && safeToRemoveOSRGuards)
+      {
+      ListIterator<TR::Block> blocksIt(&blocksInWhileLoop);
+       for (TR::Block *block = blocksIt.getCurrent(); block; block = blocksIt.getNext())
+          {
+          TR::TreeTop *lastRealTT = block->getLastRealTreeTop();
+          if (lastRealTT
+              && lastRealTT->getNode()->isOSRGuard()
+              && performTransformation(comp(), "%sRemove OSR guard n%dn from hot loop\n", OPT_DETAILS_LOOP_VERSIONER, lastRealTT->getNode()->getGlobalIndex()))
+             {
+             _cfg->removeEdge(block->getEdge(lastRealTT->getNode()->getBranchDestination()->getEnclosingBlock()));
```

Comment:
We should probably remove the the virtual guard as well using removeVirtualGuard since that is what we are doing here. We might also want to use the Block removeBranch since that would be the more official way to sort this out I think.